### ISSUE TITLE
Route53 doesn't validate TXT record lengths

### ISSUE THREAD

**Issue by wkoszek** [terraform-providers/terraform-provider-aws#14941]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform v0.13.1

### Affected Resource(s)

* aws_route53_record

### Terraform Configuration Files

```hcl

resource "aws_route53_record" "google_mail_dkim2" {
  zone_id = aws_route53_zone.segmed_ai.zone_id
  name = "google._domainkey"
  type = "TXT"
  ttl = "1799"
  records = [
    "v=DKIM1; k=rsa; p=verylongthinggoes........................................................................................................................................................................................................................................................................here"
  ]
}

```

### Panic Output

[ERR]: Error building changeset: InvalidChangeBatch: [Invalid Resource Record: FATAL problem: CharacterStringTooLong (Value is too long) encountered with '"v=DKIM1; k=rsa; p=MIIBIjANBgkqhRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRrRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRR_very_long_string"']
	status code: 400, request id: 1f3fcfb0-344a-4b66-adbd-a40788df8990

### Expected Behavior

Since 255 length limit is known, I'd expect `terraform plan` to warn me about a domain name that is too long, so that my GitOps "prep" stage could fail.

### Actual Behavior

`terraform plan` is OK with very long domains. `terraform apply` crashes.

### Steps to Reproduce

1. Take a DNS block from above. The idea is that you should attempt to make a very long (400characters+) long TXT record, which is popular in DKIM settings.
2. Run `terraform plan` and see it succeed.
3. Run `terraform apply` and see it crash with `CharacterStringTooLong` error

### References

* #0000

**Comment by wkoszek** [comment]
BTW, the solution to "fix" is to split the record to 255-byte chunks. So you end up with:

```
"1stpartof256byte\"\"other255bytepart\"
```

It's quite error prone. Great addition would be allow me to re-paste DKIM / TXT records the way GSuite / Hubspot etc software tell me too, without a need of manual tinkering.

**Comment by gmtreacy** [comment]
The records are in a schema.TypeSet and there is no helper validation support for sets.

something like this would be nice but is not possible at this time:

resource_aws_route53_record.go

```           
diff --git a/aws/resource_aws_route53_record.go b/aws/resource_aws_route53_record.go
index 63e1e26ee..daae949d3 100644
--- a/aws/resource_aws_route53_record.go
+++ b/aws/resource_aws_route53_record.go
@@ -235,6 +235,7 @@ func resourceAwsRoute53Record() *schema.Resource {
                                Elem:          &schema.Schema{Type: schema.TypeString},
                                Optional:      true,
                                Set:           schema.HashString,
+                               ValidateFunc:  validation.StringLenBetween(1, 255),
                        },
 
                        "allow_overwrite": {

```

```
zordak@sony ~/repos/issues/14941 $ terraform plan

Error: InternalValidate

Internal validation of the provider failed! This is always a bug
with the provider itself, and not a user issue. Please report
this bug:

1 error occurred:
        * resource aws_route53_record: records: ValidateFunc and ValidateDiagFunc are
not yet supported on lists or sets.

```

https://github.com/hashicorp/terraform/pull/6508
https://github.com/hashicorp/terraform-plugin-sdk/issues/156

**Comment by angelabad** [comment]
His, this is documented in resource official docs:

"records - (Required for non-alias records) A string list of records. To specify a single record value longer than 255 characters such as a TXT record for DKIM, add \"\" inside the Terraform configuration string (e.g. "first255characters\"\"morecharacters")."

**Comment by angelabad** [comment]
His, this is documented in resource official docs:

records - (Required for non-alias records) A string list of records. To specify a single record value longer than 255 characters such as a TXT record for DKIM, add \\"\\" inside the Terraform configuration string (e.g. "first255characters\\"\\"morecharacters").

**Comment by philnichol** [comment]
Hope you all don't mind but I've raised #16317 that hopefully addresses this, let me know if I've missed anything 😄

**Comment by wkoszek** [comment]
@philnichol The PR looks good, but I wonder if we're bound to the current way of doing it, just because we already have this limit in place, and for backward compatibility, or could we do something better?

If I give domain record > 255 chars, why can't AWS provider split it up for me? What's the advantage of me putting quotes in place?

**Comment by philnichol** [comment]
@wkoszek I agree with you personally. Main thought behind just adding validation for me is backward compatibility and to avoid introducing a breaking change to route53 (which is pretty
 terrifying).
Definitely one for the maintainers to advise on rather than me 😄, I'm quite new to this repo and have only contributed tiny amounts.

**Comment by zimbatm** [comment]
Hit this again today. Here is the nicest workaround I could come up with:
```tf
locals {
  dkim_record = "v=DKIM1; k=rsa; p=MIIBIjAN..."
  zone_id = "..."
  domain = "mydomain.com"
}

resource "aws_route53_record" "scarf-sh-google-domainkey" {
  zone_id = local.zone_id
  name    = "google._domainkey.${local.domain}"
  type    = "TXT"
  # AWS only accepts records up to 255 chars. We know that this one is ~410.
  # The provider doesn't know about it. So this is our workaround.
  # See https://github.com/hashicorp/terraform-provider-aws/issues/14941
  records = [
    join("\"\"", [
      substr(local.dkim_record, 0, 255),
      substr(local.dkim_record, 255, 255),
    ])
  ]
  ttl = "300"
}
```

**Comment by kevinmartin** [comment]
Another workaround, in case your record would need to be split up more than once potentially (or even 0 times, ie. within a module that creates these records). This will just insert `\"\"` after every 255th character. Works for any length of string.

```hcl
resource "aws_route53_record" "example" {
  ...
  records = [
    replace(local.dkim_record, "/(.{255})./", "$1\"\"")
  ]
}
```

**Comment by ei-grad** [comment]
Splitting a single DNS record to several DNS records automatically in `aws_route53_record` resource code is very arguable behavior. While it could be valid for DKIM records, it would lead to unexpected failures with other records. This issue only states that the validation should be in place, `terraform plan` should fail if it is clear that the record resource definition is invalid. And a TXT DNS record with length more than 255 is clearly invalid.

Was something wrong with #16317 ? Any chances such PR get merged after proper rebase and updating?

**Comment by ei-grad** [comment]
Btw, my preferred way to split string into array:
```
resource "aws_route53_record" "example" {
  ...
  records = regexall(".{1,255}", local.dkim_record)
}
```

**Comment by jimmyjayp** [comment]
not sure anyone is still looking at this, but I'm finding that with the 5.73 aws terraform provider code (terraform 1.10), the string splitting as documented for the aws_route53_record resource is not working. I've tried inserting '\"\"' in the string to effectively reduce its length and while that does indeed get past the string too long error, it also has the side effect of adding two double quotes '""' in the middle of the string invalidating the DKIM record. 

I have been unable to solve this problem having tried a number of iterations of how to chop up the string while still preserving its value as stored in the route53 zone map.

**Comment by wyardley** [comment]
Something like this also seems to work for me
```hcl
join("\"\"", [for i in range(0, length(local.input), 255) : substr(local.record, i, 255)])
```

(I remember having a similar solution in the past, but couldn't find it, and didn't love the regex based solution).
